File: server2/src/infra/repositories/in_memory_game_repository.py
from typing import List, Dict
from src.domain.games.game import Game, GameId
from src.domain.games.game_repository import GameRepository


class InMemoryGameRepository(GameRepository):

    def __init__(self):
        self.__games: Dict[GameId, Game] = {}

    def find_all_available_games(self) -> List[Game]:
        return list(filter(
            lambda game: game.is_available,
            self.__games.values()
        ))

    def find_game_by_id(self, game_id: GameId) -> Game:
        return self.__games[game_id]

    def add_game(self, game: Game):
        self.__games[game.id] = game

    def remove_game(self, game_id: GameId):
        del self.__games[game_id]

File: server-koa/src/infra/repositories/inMemoryGameRepository.ts
import { Game } from '../../domain/games/game'
import { GameId } from '../../domain/games/gameId'
import { GameRepository } from '../../domain/games/gameRepository'
import { Service } from 'typedi'

@Service()
export class InMemoryGameRepository implements GameRepository {
    private games: Map<string, Game> = new Map()

    findAllAvailable(): Array<Game> {
        return Array
            .from(this.games.values())
            .filter(game => game.isAvailable)
    }

    findById(id: GameId) {
        return this.games.get(id.toString())
    }

    save(game: Game): void {
        if (this.games.has(game.id.toString())) {
            throw new Error(`game with id '${game.id}' already exists`)
        }

        this.games.set(game.id.toString(), game)
    }
}
